# PYTHON-09 Библиотека numpy

## Модуль `collections`

### Counter

Подсчет количества вхождений каждого уникального значения в `iterable`, возвращает объект `Counter` похожий на словарь.

Импорт модуля `Counter`

```python
from collections import Counter
```

Создать пустой `Counter`:

```python
c = Counter()
```

Изменение значения отсутсвующего ключа не приводит к ошибке:

```python
c['red'] += 1
```

Для подсчета количества вхождений каждого значения в `iterable` достаточно передать его `Counter` в качестве аргумента:

```python
c = Counter(iterable)
```

При обращении к объекту `Counter` по несуществующему ключу возвращает 0, а не исключение KeyError как словарь.

Для получения всех значений используется метод объекта `Counter` - `values()`

```python
с.values()
```

Метод `values()` возвращает объект `dict_values`, который похож на список

__Из справки VS Code__
> `Counter(__iterable: None = ..., /) -> None`
Create a new, empty Counter object. And if given, count elements from an input iterable. Or, initialize the count from another mapping of elements to their counts.
```python
>>> c = Counter()                           # a new, empty counter
>>> c = Counter('gallahad')                 # a new counter from an iterable
>>> c = Counter({'a': 4, 'b': 2})           # a new counter from a mapping
>>> c = Counter(a=4, b=2)                   # a new counter from keyword args
```

In [4]:
from collections import Counter

test = ['a', 'a', 'a']
c = Counter(test)
print(c, c.values())

Counter({'a': 3}) dict_values([3])


### Операции с `Counter`

* Сложение<br>
Объекты `Counter` можно складывать. Результатом выражения будет новый объект `Counter`, в котором значения всех совпадающих ключей будут просуммированы, а несовпадающие ключи будут добавлены с имеющимися значениями.

* Вычитание
    * Метод `subtract` меняет объект `Counter` на месте, вычитая из значений его ключей, значения совпадающих ключей аргумента. Отсутствующие в изменяемом объекте ключи, будут добавлены в него с отрицательными значениями (как будто они имели в изменямом объекте значение `0`)
    * Выражения, где объекты `Counter` являются операндами оператора `-` (минус), возвращают новый объект в котором имеются только те ключи первого операнда, разность значений которых со значениями совпадающих ключей второго операнда является положительной.

In [23]:
cars_moscow = ['black', 'black', 'white', 'black', 'black', 'white', 'yellow', 'yellow', 'yellow', 'honey', 'her']
cars_spb = ['red', 'black', 'black', 'white', 'white', 'yellow', 'yellow', 'red', 'white', 'honey']

counter_moscow = Counter(cars_moscow)
counter_spb = Counter(cars_spb)
 
print(f"{counter_moscow = }")
print(f"{counter_spb = }")
print(f"{counter_moscow + counter_spb = }")
print(f"{counter_moscow - counter_spb = }")
counter_moscow.subtract(counter_spb)
print(f"After subtract {counter_moscow}")
print(f"{[*counter_spb.elements()] = }")
print(f"{counter_spb.most_common() = }")
print(f"{counter_spb.clear() = }")

counter_moscow = Counter({'black': 4, 'yellow': 3, 'white': 2, 'honey': 1, 'her': 1})
counter_spb = Counter({'white': 3, 'red': 2, 'black': 2, 'yellow': 2, 'honey': 1})
counter_moscow + counter_spb = Counter({'black': 6, 'white': 5, 'yellow': 5, 'honey': 2, 'red': 2, 'her': 1})
counter_moscow - counter_spb = Counter({'black': 2, 'yellow': 1, 'her': 1})
After subtract Counter({'black': 2, 'yellow': 1, 'her': 1, 'honey': 0, 'white': -1, 'red': -2})
[*counter_spb.elements()] = ['red', 'red', 'black', 'black', 'white', 'white', 'white', 'yellow', 'yellow', 'honey']
counter_spb.most_common() = [('white', 3), ('red', 2), ('black', 2), ('yellow', 2), ('honey', 1)]
counter_spb.clear() = None


#### Переменная

In [30]:
clients = [
    953421196, 953421161, 953421142, 953421186, 953421181, 953421144, 
    953421190, 953421184, 953421141, 953421193, 953421129, 953421158, 
    953421130, 953421177, 953421181, 953421136, 953421160, 953421184, 
    953421146, 953421175, 953421110, 953421139, 953421100, 953421116, 
    953421130, 953421179, 953421181, 953421136, 953421174, 953421167, 
    953421132, 953421195, 953421145, 953421108, 953421143, 953421133, 
    953421180, 953421149, 953421135, 953421195, 953421143, 953421131, 
    953421157, 953421189, 953421128, 953421132, 953421127, 953421151, 
    953421197, 953421160, 953421112, 953421155, 953421182, 953421168, 
    953421131, 953421156, 953421113, 953421102, 953421113, 953421192, 
    953421142, 953421105, 953421165, 953421175, 953421102, 953421195, 
    953421154, 953421165, 953421141, 953421166, 953421126, 953421143, 
    953421165, 953421150, 953421187, 953421129, 953421176, 953421169, 
    953421109, 953421177, 953421109, 953421150, 953421136, 953421140, 
    953421189, 953421198, 953421186, 953421159, 953421184, 953421182, 
    953421133, 953421103, 953421186, 953421132, 953421121, 953421107, 
    953421138, 953421190, 953421113, 953421161, 953421154, 953421161, 
    953421107, 953421113, 953421180, 953421191, 953421178, 953421116, 
    953421102, 953421167, 953421191, 953421187, 953421182, 953421118, 
    953421122, 953421157, 953421195, 953421141, 953421116, 953421176, 
    953421194, 953421163, 953421116, 953421197, 953421189, 953421177, 
    953421133, 953421138, 953421101, 953421116, 953421128, 953421104, 
    953421165, 953421113, 953421135, 953421150, 953421165, 953421154, 
    953421165, 953421161, 953421188, 953421177, 953421189, 953421128, 
    953421116, 953421177, 953421147, 953421152, 953421175, 953421151, 
    953421183, 953421195, 953421142, 953421132, 953421148, 953421112, 
    953421111, 953421121, 953421125, 953421171, 953421169, 953421127, 
    953421191, 953421194, 953421114, 953421112, 953421142, 953421144, 
    953421176, 953421120, 953421134, 953421183, 953421140, 953421144, 
    953421172, 953421189, 953421143, 953421188, 953421176, 953421142, 
    953421117, 953421140, 953421155, 953421158, 953421131, 953421187, 
    953421156, 953421120, 953421118, 953421184, 953421171, 953421186, 
    953421115, 953421101, 953421124, 953421151, 953421153, 953421141, 
    953421156, 953421144, 953421135, 953421109, 953421177, 953421144, 
    953421109, 953421148, 953421144, 953421111, 953421161, 953421133, 
    953421144, 953421118, 953421137, 953421182, 953421111, 953421178, 
    953421100, 953421160, 953421175, 953421171, 953421149, 953421156, 
    953421187, 953421198, 953421117, 953421100, 953421167, 953421118, 
    953421136, 953421101, 953421163, 953421133, 953421133, 953421136, 
    953421116, 953421141, 953421163, 953421152, 953421193, 953421197, 
    953421142, 953421172, 953421152, 953421190, 953421193, 953421102, 
    953421149, 953421117, 953421160, 953421161, 953421147, 953421170, 
    953421161, 953421147, 953421172, 953421174, 953421168, 953421121, 
    953421190, 953421162, 953421173, 953421110, 953421111, 953421154, 
    953421157, 953421161, 953421179, 953421191, 953421155, 953421139, 
    953421180, 953421189, 953421155, 953421109, 953421169, 953421174, 
    953421192, 953421104, 953421116, 953421112, 953421151, 953421169, 
    953421149, 953421176, 953421102, 953421136, 953421146, 953421152, 
    953421146, 953421152, 953421116, 953421108, 953421183, 953421128, 
    953421150, 953421158, 953421194, 953421136, 953421104, 953421139
]

### Задания 2.3-2.5

In [35]:
c = Counter(clients)
print(c.most_common())
print(c[953421102])
print(len(c))

[(953421116, 9), (953421161, 8), (953421144, 7), (953421136, 7), (953421142, 6), (953421177, 6), (953421133, 6), (953421189, 6), (953421165, 6), (953421141, 5), (953421195, 5), (953421113, 5), (953421102, 5), (953421176, 5), (953421109, 5), (953421152, 5), (953421186, 4), (953421190, 4), (953421184, 4), (953421160, 4), (953421175, 4), (953421132, 4), (953421143, 4), (953421149, 4), (953421128, 4), (953421151, 4), (953421112, 4), (953421155, 4), (953421182, 4), (953421156, 4), (953421154, 4), (953421150, 4), (953421187, 4), (953421169, 4), (953421191, 4), (953421118, 4), (953421111, 4), (953421181, 3), (953421193, 3), (953421158, 3), (953421146, 3), (953421139, 3), (953421100, 3), (953421174, 3), (953421167, 3), (953421180, 3), (953421135, 3), (953421131, 3), (953421157, 3), (953421197, 3), (953421140, 3), (953421121, 3), (953421194, 3), (953421163, 3), (953421101, 3), (953421104, 3), (953421147, 3), (953421183, 3), (953421171, 3), (953421172, 3), (953421117, 3), (953421129, 2), (953421

### `deafaultdict`

При использовании объекта `defaultdict`, в отличии от словаря, можно заранее задать тип данных, который будет автоматически создаваться, при обращении к объекту `defaultdict` по несуществующему ключу

In [3]:
from collections import defaultdict
c = defaultdict(list)
c['test'].append(1)
print(c)

defaultdict(<class 'list'>, {'test': [1]})


### `OrderedDict`

Сохраняет порядок ключей, в котором они были добавлены. Работает в том числе и на более старших версиях чем версии Python 3.7, когда порядок ключей в словаре стал языковой гарантией.

In [4]:
from collections import OrderedDict

test = [('e', 5), ('b', 4), ('c', 3), ('d', 2), ('a', 1)]

ordered_test = OrderedDict(test)
print(ordered_test)

OrderedDict([('e', 5), ('b', 4), ('c', 3), ('d', 2), ('a', 1)])


### `deque`

FIFO (от англ. First In — First Out, «первым пришёл — первым ушёл»)

LIFO (Last In — First Out, «последним пришёл — первым ушёл»).

`deque` (читается как «дек», англ. double-ended queue — двухконцевая очередь)

Методы `deque`:

*  `append` <br> добавить элемент в конец дека;
*  `appendleft` <br> добавить элемент в начало дека;
*  `pop` <br>удалить и вернуть элемент из конца дека;
*  `popleft` <br> удалить и вернуть элемент из начала дека;
*  `extend` <br> добавить итерируемый объект в конец дека;
*  `extendleft` <br> добавить итерируемый объект в начало дека; 
*  `maxlen` <br> возвращает максимальную длину дека (или `None`);
*  `reverse` <br> изменить порядок очереди на обратный;
*  `rotate`<br> перенести `n` элементов с конца в начало (при вызове с отрицательным аргументом, элементы переносятся из начала в конец)
*  `index` <br> возращает индекс первого встретившегося в последовательности элемента, переданного в качестве аргумента;
*  `count` <br> подсчитывает количество элементов равных значению, переданного при вызове.
*  `clear` <br> очистить

Объект `deque` поддерживает обращение по нижнему индексу.

Чтобы удалить элемент по нижнему индексу, необходимо воспользоваться инструкцией `del`

Именованный параметр `maxlen` задает максимальную длину дека. При добавлении новых элементов, старые будут удаляться.

### ЗАДАНИЕ 3.2

In [2]:
# Напечатайте словарь из температур, отсортированный по уменьшению температуры
from collections import OrderedDict
temps = [('2000', -4.4), ('2001', -2.5), ('2002', -4.4), ('2003', -9.5),
        ('2004', -8.2), ('2005', -1.6), ('2006', -5.9), ('2007', -2.4),
        ('2008', -1.7), ('2009', -3.5), ('2010', -12.1), ('2011', -5.8),
        ('2012', -4.9), ('2013', -6.1), ('2014', -6.9), ('2015', -2.7),
        ('2016', -11.2), ('2017', -3.9), ('2018', -2.9), ('2019', -6.5),
        ('2020', 1.5)]

temps.sort(reverse=True, key=lambda x: x[1])
print(OrderedDict(temps))

OrderedDict([('2010', -12.1), ('2016', -11.2), ('2003', -9.5), ('2004', -8.2), ('2014', -6.9), ('2019', -6.5), ('2013', -6.1), ('2006', -5.9), ('2011', -5.8), ('2012', -4.9), ('2000', -4.4), ('2002', -4.4), ('2017', -3.9), ('2009', -3.5), ('2018', -2.9), ('2015', -2.7), ('2001', -2.5), ('2007', -2.4), ('2008', -1.7), ('2005', -1.6), ('2020', 1.5)])


### Задания 3.5 - 3.7

In [58]:
from collections import deque
users = [6, 18, 4, 7, 8, 8, 5, 18, 12, 17, 13, 15, 6, 7, 9, 17, 18, 8, 4, 11, 10, 8, 2, 10, 6, 10, 10, 9]
users = deque(users)
print(users.popleft())
users.rotate(-5)
print(users[-1])
print(users.count(users.pop()))

6
8
3


### Модуль Collections. Закрепление знаний

#### ЗАДАНИЕ 4.3

Напишите функцию brackets(line), которая определяет, является ли последовательность из круглых скобок правильной.

##### Решение без стека

In [29]:
def brackets(line: str = "") -> bool:
    if len(line) % 2 != 0:
        return False
    num_brackets = 0
    for s in line:
        if s == '(':
            num_brackets += 1
        else:
            num_brackets -= 1
        if num_brackets < 0:
            return False
    return num_brackets == 0

print(brackets("(())"))

True


##### Решение со стеком (как просили в задании)

In [36]:
from collections import deque

def brackets(line: str = "") -> bool:
    brackets_deq = deque()
    for s in line:
        if s == '(':
            brackets_deq.append(1)
        elif not brackets_deq:
            return False
        else:    
            brackets_deq.pop()
    return not brackets_deq

print(brackets("(((())))"))

True


#### Задания 4.4-4.8

В переменных `center`, `south` и `north` хранятся списки из перечней купленных позиций в трёх торговых точках, расположенных в разных районах города.
Вначале избавьтесь от излишней вложенности: в каждой переменной (`center`, `south`, `north`) должен храниться объединённый список купленных товаров без разбиения по чекам.

Пример: `[['Milk', 'Bread'], ['Meat']] -> ['Milk', 'Bread', 'Meat']`

После этого определите, в каком магазине было куплено больше всего товаров.

In [17]:
north = [['Milk', 'Milk', 'Beer'], ['Chocolate', 'Bread', 'Chips'], ['Cola', 'Cola', 'Yoghurt', 'Soap', 'Beer', 'Chips', 'Milk'], ['Soap', 'Soap', 'Cola', 'Cola', 'Chips'], ['Milk', 'Beer', 'Meat', 'Ketchup', 'Cola', 'Cola', 'Chips', 'Chips', 'Cola', 'Cola'], ['Beer', 'Bread', 'Bread', 'Beer', 'Beer', 'Beer', 'Bread', 'Cheese'], ['Yoghurt', 'Beer', 'Chips', 'Milk', 'Soap', 'Cola', 'Cola', 'Cola', 'Beer', 'Cola', 'Cola', 'Cola', 'Beer', 'Ketchup', 'Beer', 'Beer', 'Beer', 'Soap'], ['Milk', 'Cola', 'Cola', 'Beer', 'Beer', 'Bread', 'Bread', 'Soap', 'Cola', 'Cola', 'Beer', 'Meat', 'Bread', 'Chips'], ['Beer', 'Beer', 'Beer', 'Chips', 'Milk', 'Cola', 'Chocolate', 'Beer', 'Chocolate', 'Beer', 'Beer', 'Cola', 'Meat', 'Yoghurt', 'Beer'], ['Bread'], ['Chocolate', 'Beer', 'Meat', 'Yoghurt'], ['Cola', 'Beer', 'Beer', 'Beer', 'Chocolate', 'Beer', 'Soap', 'Beer', 'Chips', 'Soap', 'Chocolate', 'Bread', 'Chips', 'Cola', 'Bread', 'Beer', 'Cola', 'Bread'], ['Chips', 'Cola', 'Beer', 'Chips', 'Cola', 'Cola', 'Beer', 'Soap', 'Yoghurt', 'Yoghurt', 'Cola', 'Bread', 'Beer', 'Chocolate', 'Chips', 'Bread', 'Beer', 'Bread'], ['Cola', 'Chocolate'], ['Chocolate', 'Cola', 'Meat', 'Cola', 'Ketchup', 'Cola', 'Chocolate', 'Bread', 'Chocolate', 'Chocolate', 'Meat'], ['Bread', 'Milk', 'Chips', 'Ketchup', 'Cola', 'Cola', 'Cola', 'Beer', 'Beer', 'Soap', 'Beer', 'Cola'], ['Yoghurt', 'Milk', 'Soap', 'Bread', 'Cola', 'Cola', 'Milk', 'Bread', 'Chips', 'Cheese', 'Milk', 'Yoghurt', 'Bread', 'Yoghurt'], ['Cola', 'Ketchup'], ['Cola', 'Yoghurt', 'Bread', 'Cola', 'Cola', 'Chips', 'Yoghurt', 'Milk', 'Beer', 'Chips', 'Bread', 'Beer', 'Beer', 'Cola', 'Bread', 'Beer', 'Beer', 'Cheese'], ['Beer', 'Cheese', 'Ketchup', 'Beer'], ['Beer', 'Beer', 'Beer'], ['Soap', 'Beer', 'Beer', 'Chocolate', 'Cola', 'Chocolate', 'Bread', 'Beer', 'Milk', 'Bread', 'Beer', 'Chocolate', 'Bread', 'Cola', 'Cola', 'Cheese', 'Beer', 'Cola', 'Soap', 'Yoghurt'], ['Beer', 'Chocolate'], ['Cola', 'Beer'], ['Yoghurt', 'Beer', 'Yoghurt', 'Yoghurt', 'Chips', 'Meat', 'Beer', 'Chocolate', 'Cola', 'Cola', 'Chips', 'Bread'], ['Cola', 'Cola', 'Cola', 'Cola', 'Cola', 'Bread', 'Chips', 'Soap', 'Cola', 'Chocolate', 'Beer', 'Beer'], ['Beer', 'Cola', 'Cola', 'Bread', 'Soap', 'Beer', 'Meat', 'Beer', 'Beer', 'Beer', 'Cola', 'Chips', 'Beer', 'Cola', 'Cola', 'Bread', 'Cheese', 'Beer'], ['Cola', 'Cola', 'Ketchup', 'Beer', 'Yoghurt', 'Bread'], ['Chips', 'Yoghurt', 'Cola', 'Cola', 'Cola', 'Chocolate', 'Chips', 'Bread', 'Chocolate', 'Yoghurt', 'Chocolate', 'Milk', 'Bread', 'Bread', 'Soap', 'Milk', 'Soap', 'Cola', 'Bread', 'Beer'], ['Beer', 'Beer', 'Ketchup', 'Cola', 'Beer', 'Bread', 'Beer', 'Cola', 'Beer', 'Chocolate'], ['Beer', 'Chocolate', 'Cola', 'Beer', 'Yoghurt', 'Milk', 'Bread', 'Cheese', 'Yoghurt', 'Beer', 'Cola', 'Yoghurt', 'Cola', 'Soap', 'Beer', 'Bread', 'Meat', 'Bread', 'Cola'], ['Beer', 'Cola', 'Chips', 'Cola'], ['Cola', 'Cola', 'Beer', 'Cheese'], ['Bread', 'Soap', 'Ketchup', 'Chocolate', 'Beer', 'Cola', 'Chocolate', 'Cola', 'Cola', 'Yoghurt', 'Beer', 'Bread', 'Cola', 'Ketchup', 'Beer'], ['Bread'], ['Bread', 'Beer', 'Yoghurt', 'Yoghurt', 'Bread', 'Milk', 'Soap', 'Meat', 'Bread', 'Beer', 'Cola', 'Milk', 'Milk', 'Bread', 'Beer', 'Cola', 'Ketchup', 'Cola'], ['Bread', 'Beer', 'Bread', 'Yoghurt', 'Beer', 'Bread', 'Cola', 'Cola', 'Cola', 'Beer', 'Bread', 'Milk', 'Chips', 'Cola', 'Beer', 'Bread', 'Soap', 'Bread', 'Yoghurt', 'Bread'], ['Yoghurt', 'Beer', 'Cola', 'Beer', 'Beer', 'Beer'], ['Chips', 'Chocolate', 'Soap', 'Chocolate', 'Cola', 'Bread', 'Beer', 'Cola', 'Beer', 'Ketchup', 'Chocolate', 'Ketchup', 'Ketchup', 'Cheese', 'Chips', 'Beer', 'Chips', 'Chocolate'], ['Bread', 'Cola', 'Cola', 'Beer', 'Bread', 'Bread', 'Beer', 'Chocolate', 'Bread', 'Cola', 'Milk', 'Chips', 'Meat', 'Beer', 'Beer', 'Soap', 'Bread'], ['Beer', 'Beer', 'Bread', 'Chips', 'Beer', 'Bread', 'Bread', 'Chips', 'Beer'], ['Yoghurt', 'Bread', 'Cola', 'Bread', 'Cola', 'Bread', 'Meat', 'Cola', 'Bread', 'Beer', 'Soap', 'Chips'], ['Bread', 'Beer'], ['Milk', 'Beer', 'Meat', 'Cola', 'Beer', 'Cola', 'Cola'], ['Beer', 'Chips', 'Yoghurt', 'Beer', 'Cola', 'Beer', 'Cola', 'Cola', 'Soap', 'Cola'], ['Bread', 'Cola', 'Cola', 'Meat'], ['Cola', 'Chocolate', 'Meat', 'Beer', 'Cola', 'Bread', 'Chips', 'Beer', 'Chips', 'Chips', 'Cola'], ['Bread', 'Cola', 'Cola', 'Cola', 'Beer', 'Cola', 'Yoghurt', 'Beer', 'Chips', 'Cola', 'Chocolate', 'Chips', 'Cola', 'Cola', 'Cola', 'Cola', 'Bread', 'Cola'], ['Cola', 'Soap', 'Cola'], ['Soap', 'Chips', 'Cola', 'Beer', 'Bread', 'Soap', 'Cheese', 'Bread', 'Beer', 'Chocolate']]
center = [['Meat', 'Beer', 'Soap', 'Beer', 'Cheese', 'Cola', 'Milk', 'Soap', 'Cola', 'Meat', 'Bread', 'Chocolate', 'Chips'], ['Soap', 'Beer', 'Chips', 'Bread', 'Beer', 'Beer', 'Beer', 'Cheese', 'Cheese', 'Beer', 'Chips', 'Chocolate', 'Chips', 'Cheese', 'Bread', 'Cola', 'Cola', 'Beer'], ['Cola', 'Soap', 'Bread', 'Milk', 'Beer', 'Meat', 'Bread', 'Bread'], ['Ketchup', 'Beer', 'Ketchup', 'Chocolate', 'Milk', 'Milk', 'Bread', 'Beer'], ['Beer', 'Beer', 'Meat', 'Ketchup', 'Soap', 'Bread', 'Cola', 'Beer'], ['Meat', 'Bread', 'Milk', 'Cheese', 'Soap', 'Beer', 'Milk', 'Cheese', 'Cola', 'Beer', 'Chips', 'Bread', 'Ketchup', 'Chocolate', 'Bread', 'Milk'], ['Yoghurt'], ['Beer', 'Milk', 'Chips', 'Soap', 'Chips', 'Milk', 'Beer', 'Chips', 'Bread', 'Meat', 'Milk'], ['Yoghurt', 'Beer', 'Cola', 'Cola', 'Beer', 'Soap', 'Cheese', 'Soap', 'Bread', 'Cola', 'Yoghurt', 'Ketchup', 'Beer', 'Milk'], ['Milk', 'Cola', 'Bread', 'Cola', 'Bread', 'Beer', 'Beer', 'Beer'], ['Yoghurt', 'Cola'], ['Bread', 'Yoghurt', 'Chips', 'Ketchup', 'Meat', 'Bread', 'Beer', 'Yoghurt', 'Cola', 'Cola'], ['Chips', 'Chocolate', 'Chips', 'Meat', 'Bread', 'Cheese', 'Bread', 'Yoghurt'], ['Ketchup', 'Soap', 'Chocolate', 'Bread'], ['Chips', 'Beer', 'Chips', 'Cola', 'Cheese', 'Soap', 'Ketchup', 'Meat', 'Cola', 'Chips', 'Beer', 'Chocolate', 'Beer', 'Milk', 'Bread', 'Ketchup', 'Chips', 'Cheese', 'Ketchup'], ['Beer', 'Milk', 'Soap', 'Chips', 'Soap', 'Bread', 'Bread', 'Milk', 'Beer'], ['Cola', 'Chips', 'Meat', 'Cola', 'Beer', 'Chocolate', 'Bread', 'Bread', 'Chips', 'Soap', 'Chocolate', 'Chips', 'Beer'], ['Meat', 'Cola', 'Chips', 'Bread', 'Chips', 'Chocolate', 'Bread', 'Meat', 'Bread', 'Yoghurt', 'Cheese', 'Bread', 'Chips', 'Cola'], ['Chips', 'Cheese', 'Bread', 'Beer', 'Bread', 'Chips', 'Chocolate', 'Bread', 'Cola', 'Cola', 'Chocolate', 'Chocolate', 'Bread', 'Meat', 'Chips'], ['Bread', 'Milk', 'Bread', 'Cheese', 'Bread', 'Cheese', 'Ketchup', 'Beer', 'Cheese', 'Cola', 'Milk', 'Milk', 'Bread', 'Beer', 'Bread', 'Chips'], ['Yoghurt'], ['Bread', 'Bread', 'Chips', 'Cheese', 'Bread', 'Beer', 'Cola', 'Ketchup', 'Bread', 'Chips', 'Chocolate', 'Meat', 'Milk', 'Beer', 'Milk', 'Cheese', 'Bread', 'Meat', 'Bread', 'Cola'], ['Bread', 'Meat', 'Meat', 'Milk', 'Chips', 'Soap', 'Yoghurt', 'Chips', 'Beer', 'Yoghurt'], ['Bread', 'Soap', 'Bread', 'Cola', 'Bread'], ['Cola', 'Bread', 'Meat', 'Cola', 'Meat', 'Chocolate', 'Chips', 'Meat', 'Chips'], ['Chips', 'Cheese', 'Cheese', 'Meat'], ['Chips', 'Meat', 'Soap', 'Cheese', 'Bread', 'Cola', 'Bread', 'Beer', 'Meat', 'Cola', 'Bread', 'Cola', 'Ketchup', 'Bread'], ['Chips', 'Cheese', 'Milk', 'Meat', 'Milk', 'Beer', 'Chocolate', 'Ketchup', 'Cola', 'Cheese', 'Beer'], ['Beer', 'Ketchup', 'Yoghurt', 'Ketchup', 'Chocolate', 'Bread', 'Beer', 'Ketchup', 'Chocolate', 'Cola', 'Chocolate', 'Ketchup', 'Cola', 'Meat', 'Chips', 'Soap', 'Meat'], ['Meat', 'Milk'], ['Cola', 'Beer', 'Yoghurt', 'Beer', 'Bread', 'Cola'], ['Chips', 'Meat', 'Cheese', 'Ketchup', 'Chips', 'Bread', 'Bread', 'Chips', 'Chips', 'Bread', 'Milk', 'Ketchup', 'Cola', 'Cola', 'Beer'], ['Beer', 'Bread', 'Cheese', 'Bread', 'Cola', 'Cheese', 'Cheese', 'Beer', 'Milk', 'Bread', 'Chocolate', 'Cheese', 'Beer', 'Bread', 'Beer', 'Cola', 'Yoghurt', 'Beer', 'Beer', 'Chips'], ['Bread', 'Chips', 'Bread', 'Cola', 'Chips', 'Chocolate', 'Cheese', 'Beer', 'Chips', 'Milk', 'Milk', 'Beer', 'Cola', 'Meat', 'Cola', 'Bread', 'Cola', 'Chocolate', 'Chocolate', 'Cola'], ['Soap', 'Yoghurt', 'Chips', 'Beer', 'Chips', 'Milk', 'Cheese', 'Meat', 'Beer', 'Bread', 'Ketchup', 'Bread', 'Bread', 'Cheese', 'Milk', 'Beer', 'Beer', 'Soap', 'Bread'], ['Cola', 'Bread', 'Cheese', 'Ketchup', 'Beer', 'Chips', 'Meat', 'Chocolate', 'Chips', 'Cola', 'Beer', 'Beer', 'Cola'], ['Ketchup', 'Beer', 'Chocolate', 'Bread', 'Yoghurt', 'Beer', 'Cheese'], ['Bread', 'Chocolate', 'Bread', 'Milk'], ['Meat', 'Yoghurt', 'Bread', 'Yoghurt', 'Cola', 'Ketchup'], ['Ketchup', 'Bread', 'Bread', 'Chocolate', 'Chocolate', 'Chocolate', 'Bread', 'Bread', 'Beer', 'Chocolate', 'Bread', 'Milk'], ['Bread', 'Cheese', 'Soap', 'Soap', 'Chips', 'Chips'], ['Ketchup', 'Chocolate', 'Chips', 'Milk', 'Soap', 'Soap', 'Ketchup', 'Bread', 'Ketchup', 'Cola', 'Cheese', 'Beer', 'Ketchup', 'Bread'], ['Bread', 'Milk', 'Beer', 'Yoghurt', 'Meat', 'Ketchup', 'Meat', 'Meat', 'Bread', 'Milk', 'Cheese', 'Beer', 'Yoghurt', 'Milk', 'Bread', 'Cola'], ['Chips', 'Cola', 'Milk', 'Chocolate', 'Beer', 'Beer', 'Chips', 'Bread', 'Beer', 'Beer', 'Bread', 'Beer', 'Ketchup', 'Milk', 'Yoghurt', 'Ketchup', 'Cola', 'Ketchup', 'Chips', 'Meat'], ['Beer', 'Bread', 'Soap', 'Cheese', 'Meat', 'Soap'], ['Beer', 'Meat', 'Beer', 'Yoghurt', 'Soap', 'Chips', 'Meat', 'Cheese', 'Milk', 'Bread', 'Meat', 'Beer', 'Milk'], ['Chips', 'Meat', 'Bread'], ['Chocolate', 'Soap', 'Bread', 'Chips', 'Chips', 'Yoghurt', 'Chips', 'Cola', 'Cola', 'Cola', 'Beer', 'Milk', 'Milk', 'Bread', 'Bread', 'Meat'], ['Meat', 'Chocolate', 'Chips', 'Chips', 'Yoghurt', 'Yoghurt', 'Beer', 'Cola', 'Cheese', 'Milk', 'Beer'], ['Meat', 'Chocolate', 'Yoghurt', 'Cola', 'Cheese', 'Meat', 'Bread', 'Beer', 'Meat', 'Beer', 'Yoghurt', 'Cola', 'Bread']]
south = [['Cola', 'Meat', 'Cheese', 'Yoghurt', 'Beer', 'Milk', 'Milk', 'Meat', 'Cola', 'Cola', 'Cheese', 'Beer', 'Yoghurt', 'Beer', 'Bread', 'Bread', 'Milk', 'Cheese', 'Chocolate'], ['Soap', 'Milk', 'Cola'], ['Milk', 'Bread', 'Yoghurt', 'Meat', 'Meat'], ['Bread', 'Milk', 'Beer'], ['Beer'], ['Chocolate', 'Meat', 'Chocolate', 'Cola', 'Cola', 'Cola', 'Cola', 'Yoghurt', 'Bread', 'Meat', 'Soap', 'Soap', 'Milk', 'Milk', 'Cola'], ['Beer', 'Beer', 'Meat', 'Chips', 'Bread', 'Bread', 'Bread', 'Bread', 'Milk', 'Cola', 'Chocolate', 'Bread', 'Beer', 'Chips', 'Bread', 'Bread', 'Yoghurt'], ['Chips', 'Milk', 'Soap'], ['Meat', 'Beer', 'Milk', 'Chocolate', 'Bread', 'Yoghurt'], ['Chips', 'Meat', 'Chocolate', 'Bread', 'Cola', 'Cola', 'Chocolate', 'Meat', 'Yoghurt', 'Milk'], ['Bread', 'Soap', 'Bread', 'Meat', 'Beer', 'Yoghurt', 'Milk', 'Cola', 'Bread', 'Ketchup'], ['Meat', 'Milk'], ['Meat', 'Beer', 'Yoghurt'], ['Cola', 'Bread', 'Cola', 'Chocolate', 'Chips', 'Meat', 'Cheese'], ['Milk', 'Milk', 'Cheese', 'Meat'], ['Chips', 'Yoghurt', 'Cheese', 'Soap', 'Ketchup', 'Cheese', 'Soap', 'Beer', 'Ketchup', 'Ketchup', 'Milk', 'Bread', 'Bread', 'Beer'], ['Meat'], ['Ketchup', 'Bread', 'Beer', 'Milk', 'Bread', 'Meat', 'Ketchup', 'Cheese'], ['Meat', 'Chips', 'Bread', 'Meat', 'Milk', 'Soap', 'Chocolate', 'Meat', 'Chocolate', 'Chocolate', 'Bread', 'Cheese', 'Soap', 'Cola', 'Yoghurt'], ['Cheese', 'Milk', 'Bread', 'Milk', 'Chips', 'Chips', 'Meat', 'Beer', 'Chocolate', 'Chocolate'], ['Ketchup', 'Beer', 'Cheese', 'Cola'], ['Chocolate', 'Cheese', 'Bread'], ['Milk', 'Yoghurt', 'Ketchup', 'Beer', 'Meat', 'Chips', 'Yoghurt', 'Meat', 'Bread', 'Chips'], ['Yoghurt', 'Milk', 'Ketchup', 'Yoghurt', 'Beer', 'Cheese', 'Bread', 'Bread', 'Ketchup', 'Bread', 'Bread', 'Yoghurt', 'Meat'], ['Soap', 'Meat', 'Bread', 'Beer', 'Milk'], ['Beer', 'Cola', 'Beer', 'Meat', 'Meat', 'Cheese', 'Meat', 'Chocolate', 'Bread', 'Ketchup', 'Milk', 'Soap'], ['Cheese', 'Chocolate', 'Milk', 'Chocolate', 'Cola', 'Bread', 'Chips', 'Cheese', 'Soap', 'Ketchup', 'Cheese', 'Chips', 'Cheese', 'Cola', 'Chocolate', 'Beer'], ['Bread', 'Bread', 'Cola', 'Ketchup', 'Cola', 'Bread', 'Meat', 'Yoghurt', 'Milk', 'Beer', 'Beer', 'Cheese', 'Meat', 'Bread', 'Cheese', 'Meat', 'Chocolate'], ['Chocolate', 'Soap', 'Chips', 'Beer', 'Bread', 'Yoghurt', 'Chips', 'Chocolate', 'Beer', 'Cheese', 'Cola', 'Milk', 'Chips', 'Milk', 'Ketchup', 'Cola', 'Meat', 'Beer', 'Cheese', 'Yoghurt'], ['Soap'], ['Meat', 'Beer', 'Milk', 'Bread', 'Beer', 'Cheese', 'Chocolate', 'Beer', 'Beer', 'Milk', 'Beer', 'Bread', 'Meat', 'Beer', 'Chocolate', 'Beer', 'Soap', 'Chips', 'Cola'], ['Cola', 'Beer', 'Meat', 'Chips', 'Soap', 'Cola', 'Bread', 'Cola', 'Bread', 'Chips', 'Ketchup', 'Ketchup', 'Beer', 'Ketchup', 'Cola', 'Milk', 'Cheese'], ['Cheese', 'Milk', 'Chips', 'Bread', 'Yoghurt', 'Soap', 'Beer', 'Chips', 'Ketchup', 'Chips', 'Beer', 'Yoghurt', 'Cola', 'Cheese', 'Chocolate', 'Beer'], ['Meat', 'Bread', 'Meat', 'Bread'], ['Cola', 'Beer', 'Yoghurt'], ['Beer', 'Bread', 'Beer', 'Meat', 'Bread', 'Milk', 'Soap', 'Milk', 'Chocolate', 'Meat', 'Meat', 'Meat', 'Chips', 'Chocolate', 'Meat'], ['Beer', 'Cola', 'Chocolate', 'Bread', 'Cheese', 'Cheese'], ['Milk', 'Chips', 'Cola', 'Milk', 'Bread', 'Bread', 'Beer', 'Milk', 'Cola', 'Chocolate', 'Chocolate', 'Meat', 'Cola', 'Cola', 'Beer', 'Cola', 'Chocolate', 'Bread', 'Bread', 'Cola'], ['Chocolate', 'Chocolate', 'Beer', 'Beer', 'Bread', 'Yoghurt', 'Meat', 'Cola', 'Yoghurt'], ['Ketchup', 'Cola', 'Ketchup', 'Yoghurt', 'Chips', 'Soap', 'Soap', 'Chocolate', 'Chocolate', 'Bread', 'Beer', 'Meat'], ['Bread', 'Meat', 'Soap', 'Cola', 'Bread', 'Cola', 'Yoghurt', 'Meat', 'Bread', 'Cola', 'Cola', 'Ketchup', 'Beer', 'Bread', 'Milk', 'Yoghurt', 'Meat'], ['Chocolate', 'Yoghurt', 'Bread'], ['Meat', 'Bread', 'Bread', 'Bread'], ['Beer', 'Milk', 'Cola', 'Ketchup', 'Cola', 'Cheese', 'Meat', 'Chocolate'], ['Soap', 'Beer', 'Chocolate', 'Chocolate', 'Cola', 'Cola', 'Yoghurt', 'Ketchup', 'Milk'], ['Meat', 'Yoghurt', 'Bread', 'Ketchup', 'Ketchup', 'Milk', 'Meat'], ['Ketchup', 'Soap', 'Chips', 'Ketchup', 'Bread', 'Chocolate', 'Milk', 'Bread', 'Bread', 'Ketchup', 'Cola', 'Meat', 'Milk', 'Bread', 'Cola'], ['Meat', 'Beer', 'Yoghurt', 'Chips', 'Beer', 'Meat', 'Cola', 'Beer', 'Meat', 'Ketchup', 'Milk', 'Cola', 'Yoghurt', 'Beer', 'Meat', 'Bread', 'Bread'], ['Meat', 'Soap', 'Cheese', 'Ketchup', 'Cola', 'Cola', 'Bread', 'Chips', 'Meat', 'Cola', 'Bread', 'Beer', 'Beer', 'Beer'], ['Meat', 'Yoghurt', 'Bread', 'Milk', 'Beer', 'Beer']]

In [89]:
"""Tasks 4.4-4.7 from PYTHON-09"""

from collections import Counter, defaultdict

all_stores = {
    'north': Counter([grocery for check in north for grocery in check]),
    'center':Counter([grocery for check in center for grocery in check]),
    'south': Counter([grocery for check in south for grocery in check])
}

# 4.4 В каком магазине было куплено больше всего товаров.
all_shops_num_sales = {
    shop: sum(number.values()) for shop, number in all_stores.items()
    }
print("The store with the biggest sales is",
      max(all_shops_num_sales, key=all_shops_num_sales.get))

# 4.5 Сколько раз покупали самый редкий товар в магазине north?
# Запишите ответ в числовой форме.
print("Number of sales of the rarest product in the north store is",
      min(all_stores['north'].values()))

# 4.6 Выберите товар, который в магазине center покупали чаще,
# чем в магазине north:
search_product = ['Beer', 'Cola', 'Bread', 'Yoghurt']
for product in search_product:
    if all_stores['center'][product] > all_stores['north'][product]:
        print("The product that was saled in the center "
              "store more often than in the north store is", product)

# 4.7 Есть ли такой товар, который в одном из магазинов покупали чаще,
# чем в двух других вместе взятых?
# Если да, выберите магазин с настолько популярным товаром:
all_products = defaultdict(dict)
for store, products in all_stores.items():
    for product, number in products.items():
        all_products[product][store] = number

for product, stores in all_products.items():
    if max(stores.values()) > (sum(stores.values()) - max(stores.values())):
        print("The product that was bought in one store more often"
              " than in the other two combined is", product, stores)

# 4.8 Определите суммарное число продаж каждого товара во всех магазинах,
# сложив все объекты-счётчики.
# Сколько раз был куплен второй по популярности товар?
# Запишите ответ в числовой форме.
sum_products = Counter()
sum_products = sum(all_stores.values(), sum_products)
print("Most popular products:")
print(sum_products)


The store with the biggest sales is center
Number of sales of the rarest product in the north store is 10
The product that was saled in the center store more often than in the north store is Bread
The product that was bought in one store more often than in the other two combined is Cola {'north': 121, 'center': 60, 'south': 55}
Most popular products:
Counter({'Beer': 257, 'Bread': 240, 'Cola': 236, 'Chips': 128, 'Meat': 121, 'Milk': 111, 'Chocolate': 107, 'Yoghurt': 90, 'Soap': 80, 'Cheese': 79, 'Ketchup': 77})


#### Задание 4.9

Дан список кортежей ratings с рейтингами кафе. Кортеж состоит из названия
и рейтинга кафе.

Необходимо отсортировать список кортежей по убыванию рейтинга. Если рейтинги
совпадают, то отсортировать кафе дополнительно по названию в алфавитном
порядке.

Получите словарь cafes с упорядоченными ключами из отсортированного списка, 
где ключи — названия кафе, а значения — их рейтинг.

In [2]:
ratings = [('Old York', 3.3), ('New Age', 4.6), ('Old Gold', 3.3), ('General Foods', 4.8),
           ('Belissimo', 4.5), ('CakeAndCoffee', 4.2), ('CakeOClock', 4.2), ('CakeTime', 4.1),
           ('WokToWork', 4.9), ('WokAndRice', 4.9), ('Old Wine Cellar', 3.3), ('Nice Cakes', 3.9)]

In [3]:
"""Task 4.9 from PYTHON-09"""

from collections import OrderedDict

ratings.sort(key=lambda x: (-x[1], x[0]))
cafes = OrderedDict(ratings)

print(cafes)


OrderedDict([('WokAndRice', 4.9), ('WokToWork', 4.9), ('General Foods', 4.8), ('New Age', 4.6), ('Belissimo', 4.5), ('CakeAndCoffee', 4.2), ('CakeOClock', 4.2), ('CakeTime', 4.1), ('Nice Cakes', 3.9), ('Old Gold', 3.3), ('Old Wine Cellar', 3.3), ('Old York', 3.3)])


#### Задание 4.10

Напишите функцию task_manager, которая принимает список задач <br>для нескольких серверов. Каждый элемент списка состоит из кортежа (<номер задачи>, <название <br>сервера>, <высокий приоритет задачи>).

Функция должна создавать словарь и заполнять его задачами <br>по следующему принципу: название сервера — ключ, по которому хранится очередь задач для конкретного <br>сервера. Если поступает задача без высокого приоритета <br>(последний элемент кортежа — False), добавить номер задачи в конец очереди. Если приоритет <br>высокий, добавить номер в начало.

Для словаря используйте defaultdict, для очереди — deque.

Функция возвращает полученный словарь с задачами.

In [4]:
tasks = [(36871, 'office', False),
(40690, 'office', False),
(35364, 'voltage', False),
(41667, 'voltage', True),
(33850, 'office', False)]

In [5]:
from collections import defaultdict, deque

def task_manager(task_list: list) -> defaultdict:
    """Function returns queues for each server.

    Args:
        task_list (list): includes corteges
                        (<number of task>,
                        <name of server>,
                        <task prioity: False or True>)

    Returns:
        defaultdict: key - name of server,
                     value - deque of server's queue
    """
    
    result = defaultdict(deque)
    for task in task_list:
        if task[2]:
            result[task[1]].appendleft(task[0])
        else:
            result[task[1]].append(task[0])
    return result

print(task_manager(tasks))

defaultdict(<class 'collections.deque'>, {'office': deque([36871, 40690, 33850]), 'voltage': deque([41667, 35364])})


## Модуль `numpy`

### Типы данных

Общепринятое сокращение `np`

```python
import numpy as np
```

Цифра в конце идентификатора функции означает сколько бит памяти будет выделено <br>для хранения переменной (по умолчанинию 64 бита)
, и, соответственно, какой доступен<br> диапазон значений для переменной 
```python
a = np.int8(100)
```

Функция `iinfo` показывает диапазон значений для конкретной переменной или для типа

```python
np.iinfo(np.int8)
np.iinfo(a)
```

Беззнаковые целочисленные типы данных (unsigned)
```python
b = np.uint8(124)
```

Числа с плавающей точкой (метод `float16`, `float32`, `float64`, `float128`)

Функция `finfo` показывает точность и диапазон значений для каждого типа


`np.finfo(np.float16)`

`# finfo(resolution=0.001, min=-6.55040e+04, max=6.55040e+04, dtype=float16)`

`np.finfo(np.float32)`

`# finfo(resolution=1e-06, min=-3.4028235e+38, max=3.4028235e+38, dtype=float32)`

`np.finfo(np.float64)`

`# finfo(resolution=1e-15, min=-1.7976931348623157e+308, max=1.7976931348623157e+308, dtype=float64)`

`np.finfo(np.float128)`

`# finfo(resolution=1e-18, min=-1.189731495357231765e+4932, max=1.189731495357231765e+4932, dtype=float128)`


Функция `sctypeDict` выводит все доступные типы. (словарь, в котором более 100 ключей)

```python
print(*sorted(map(str, set(np.sctypeDict.values()))), sep='\n')
```
Имеются также типы для работы с датой и временем, строковый тип (`str_`) и булевый тип (`bool_`)

In [112]:
print(*sorted(map(str, set(np.sctypeDict.values()))), sep='\n')

<class 'numpy.bool_'>
<class 'numpy.bytes_'>
<class 'numpy.complex128'>
<class 'numpy.complex256'>
<class 'numpy.complex64'>
<class 'numpy.datetime64'>
<class 'numpy.float128'>
<class 'numpy.float16'>
<class 'numpy.float32'>
<class 'numpy.float64'>
<class 'numpy.int16'>
<class 'numpy.int32'>
<class 'numpy.int64'>
<class 'numpy.int8'>
<class 'numpy.longlong'>
<class 'numpy.object_'>
<class 'numpy.str_'>
<class 'numpy.timedelta64'>
<class 'numpy.uint16'>
<class 'numpy.uint32'>
<class 'numpy.uint64'>
<class 'numpy.uint8'>
<class 'numpy.ulonglong'>
<class 'numpy.void'>


### Приведение типов

1. В инструкции присваивания надо указывать нужный тип данных `numpy`.
2. Арифметические операции сохраняют тип данных `numpy`.
3. Если арифметическая операция проводится с двумя разными `numpy` типами, <br>результат сохраняется с более "старшим" типом (т.е. тем, на который выделено больше памяти)
4. Если присваиваемое переменной значение превосходит доступный диапазон, то никаких ошибок <br>не вызывается, в качестве результата в переменную запишется остаток от деления <br>на максимальное доступное значение. 
5. При возникновении переполнения в результате арифметической операции <br>выводится предупреждение.

### Массивы

`import numpy as np` - `np` - общепринятие сокращение

`np.array` - создать массив

`np.array(object, dtype=None, *, copy=True, order='K', subok=False, ndmin=0, like=None)`

Именнованный параметр `dtype` задает тип данных для создаваемого массива.

`np.nan` - несуществующее число (аналог `None`)

#### Получение элемента по нижнему индексу

`<array>[ось0, ось1, ...]`

В любой из осей можно использовать срез. Если значение опущено возвращаются все элементы. 

Вместо пропущенных осей можно использовать `...`

также поддерживается индексация при помощи массивов целых чисел и булевых переменных.

#### Атрибуты массива:

- `.dtype` - тип данных, которые хранятся в массиве.

- `.ndim` - размерность (количество осей индексации) массива. 

- `.size` - общее количество элементов в массиве.

- `.shape` - кортеж, в котором содержатся длины каждой из размерностей массива (__можно менять форму массива!!!__).

- `.itemsize` - число байт, которое занимает каждый из элементов массива.


`np.zeros(shape, dtype=float, order='C', *, like=None)` - создать массив нужной формы из нулей, можно указать тип данных.

`np.arange([start,] stop[, step,], dtype=None, *, like=None)` - создать массив из диапазона.


#### Функция `np.linspace` - создать массив из диапазона чисел с заданным количеством элементов.

`np.linspace(start: _ArrayLikeFloat_co, stop: _ArrayLikeFloat_co, num: SupportsIndex = ..., endpoint: bool = ..., retstep: Literal[False] = ..., dtype: None = ..., axis: SupportsIndex = ...) -> NDArray[floating]`

- `start` - начальное значение последовательности

- `stop` - конечное значение последовательности (может включаться в нее или не включаться)

- `num` - (опционально) количество элементов последовательности (по умолчанию `50`)

- `endpoint` - (опционально) по умолчанию `True`, указывает надо ли включать `stop` в последовательность.

- `retstep` - (опционально) если `True` - функция возвращает (последовательность, шаг)

- `dtype` - (опционально) тип возврашаемого массива. Если не задан выводится из значений `start` и `stop`, но всегда `float`

- `axis` - (опционально) Ось в возвращаемом массиве для хранения сгенерированных чисел. <br>Актуально только в том случае, если `start` или `stop` являются массивоподобными.<br> По умолчанию (0) выборки будут располагаться вдоль новой оси, вставленной<br> в начале. Используйте -1, чтобы получить ось в конце.


#### Метод `.reshape` 

Возвращает копию массива нужной формы

`<array>.reshape(shape: _ShapeLike, /, *, order: _OrderACF = ...) -> ndarray[Any, dtype[float128]]`

- `order = 'C'` - (по умолчанию) массив заполняется по строкам

- `order = 'F'` - массив заполняется по столбцам

#### Метод `.transpose`

Транспонировать массив (поменять местами строки и столбцы)

`<array>.transpose(axes: _ShapeLike | None, /) -> NDArray[float128]`

#### Функция `isnan`

Возвращает массив булевых переменных, который получается в результате соспоставления <br>
каждого элемента исходного массива с `nan`

### Функция `sqrt`

`np.sqrt(__x1: _ScalarLike_co, /, out: None = ..., *, where: _ArrayLikeBool_co | None = ..., casting: _CastingKind = ..., order: _OrderKACF = ..., dtype: DTypeLike = ..., subok: bool = ..., signature: str | _2Tuple[str | None] = ..., extobj: list = ...) -> Any`

Извлечь корень из каждого элемента массива.



### Векторы

__Вектор__ - одномерный массив

При применении арифметических или логических операций между <br> 
одномерными массивами `numpy`, эта операция будет применена к каждой<br>
паре элементов массивов.<br>
Если их длина отличается, будет вызвано исключение `ValueError`

При применении арифметической операции между одномерным массивом и числом,<br>
эта операция будет произведена между каждым его элементом и числов.

#### `np.linalg` 

Модуль `numpy`, содержащий функции линейной алгебры

__Длина вектора `x`__ - это квадратный корень из суммы квадратов его координат

`length_x = np.sqrt(np.sum(x ** 2))`

__Расстояние между векторами `x` и `y`__ - квадратный корень из суммы квадратов<br>
разностей соответствующих координат.

`distance = np.sqrt(np.sum((x - y) ** 2))`

#### Функция `np.linalg.norm()`

`np.linalg.norm()` - возвращает длину вектора 

`np.linalg.norm(x - y)` - возвращает расстояние между векторами `x` и `y`


__Скалярное произведение двух векторов `x`, `y`__

`np.sum(x * y)`

`np.dot(x, y)` - если скалярное произведение равно `0`, то вектора перпендикулярны друг другу.

### Статистические функции `numpy`

[Ссылка на стат. функции `numpy`](https://pyprog.pro/statistics_functions/statistics_function.html)

`np.mean` - вычислить среднее

`np.min` - получить мимнимум

`np.max` - получить максимум

### Случайные числа

Для генерации псевдослучайных чисел используются функции модуля `random`

`rand()` 

генерация числа от 0 до 1 (1 не включая)

В качестве аргументов можно указать форму (размерности) массива в виде<br>
аргументов через запятую. Тогда функция вернет массив случайных чисел.

`sample((<форма массива>))`

Создает массив случайных чисел нужной размерности.<br>
В качестве аргумента принимает кортеж.

`uniform(low=0.0, high=1.0, size=None)` 

Возвращает случайное число из заданного диапазона.<br>
Может создавать массивы нужного размера<br>
(размеры передаются как кортеж для параметра `size`)

`randint(low, high=None, size=None, dtype=int)`

Возвращает целое случайное число.<br>
Если `high` не задано, число возрващает от `0` до `low - 1`<br>
Остальные аргументы стандартные.

`shuffle()`

Перемешивает массив на месте

`permutation()`

Возвращает копию перемешанного массива (или любой<br>
последовательности, которую преобразовывает в массив).<br>
Если в качестве аргумента указано целое число `n`,<br>
возвращает перемешанную последовательность от 0 до `n - 1`.

`choice(a, size=None, replace=True)`

Получить случайный набор объектов из массива.

- `a` — одномерный массив или число для генерации `arange(a)`;
- `size` — желаемая форма массива (число для получения<br>
одномерного массива, кортеж — для многомерного; если<br>
параметр не задан, возвращается один объект);
- `replace` — параметр, задающий, могут ли элементы повторяться<br>
(по умолчанию могут).

`seed()`

Получить одинаковые (воспроизводимые) последовательности случайных чисел.<br>
аргумент может быть от $0$ до $2^{32} - 1$

## Закрепление знаний

### Задание 10.6

Напишите функцию get_chess, которая принимает на вход длину стороны квадрата a и возвращает двумерный массив формы (a, a), заполненный 0 и 1 в шахматном порядке. В левом верхнем углу всегда должен быть ноль.

In [31]:
import numpy as np

def get_chess(side: int) -> np.ndarray:
    """Gets the length of the side of the chessboard and returns the whole board
    filled with 0 and 1 (starting from 0)
    
    Args:
        side (int): side of chessboard

    Returns:
        np.ndarray: chessboard
    """
    
    chessboard = np.zeros((side, side), dtype=np.int8)
    chessboard[::2, 1::2] = 1
    chessboard[1::2, ::2] = 1
    print(type(chessboard))
    return chessboard


print(get_chess(9))

<class 'numpy.ndarray'>
[[0 1 0 1 0 1 0 1 0]
 [1 0 1 0 1 0 1 0 1]
 [0 1 0 1 0 1 0 1 0]
 [1 0 1 0 1 0 1 0 1]
 [0 1 0 1 0 1 0 1 0]
 [1 0 1 0 1 0 1 0 1]
 [0 1 0 1 0 1 0 1 0]
 [1 0 1 0 1 0 1 0 1]
 [0 1 0 1 0 1 0 1 0]]


### Задание 10.7

Вы разрабатываете приложение для прослушивания музыки. Конечно же, там будет<br> 
доступна функция перемешивания плейлиста. Пользователю может настолько понравиться<br>
перемешанная версия плейлиста, что он захочет сохранить его копию. Однако вы <br>
не хотите хранить в памяти новую версию плейлиста, а просто хотите сохранять тот `seed`,<br>
с которым он был сгенерирован.

Для этого напишите функцию `shuffle_seed(<array>)`,  которая принимает на вход массив<br>
из чисел, генерирует случайное число для `seed` в диапазоне от $0$ до $2^{32} - 1$ (включительно)<br>
и возвращает кортеж: перемешанный с данным `seed` массив (исходный массив должен оставаться <br>
без изменений), а также `seed`, с которым этот массив был получен.

Пример:

```python
array = [1, 2, 3, 4, 5]
shuffle_seed(array)
# (array([1, 3, 2, 4, 5]), 2332342819)
shuffle_seed(array)
# (array([4, 5, 2, 3, 1]), 4155165971)
```

In [24]:
import numpy as np

def shuffle_seed(playlist: np.ndarray) -> tuple:
    """Generates and returns a random integer and shuffled
    playlist by this integer as seed.

    Args:
        playlist (np.ndarray): playist for shuffle
    Returns:
        tuple: playlist(np.ndarray), num_for_seed(np.uint32) 
        
    """

    num_for_seed = np.random.randint(0, np.iinfo(np.uint32).max, dtype=np.uint32)
    np.random.seed(num_for_seed)
    return np.random.permutation(playlist), num_for_seed
    
    
array = [1, 2, 3, 4, 5]
shuffle_seed(array)

(array([2, 3, 1, 5, 4]), 2239938729)

### Задание 10.8

Напишите функцию `min_max_dist`, которая принимает на вход неограниченное<br>
число векторов через запятую. Гарантируется, что все векторы, которые<br>
передаются, одинаковой длины.

Функция возвращает минимальное и максимальное расстояние между векторами<br>
в виде кортежа.

Пример:
```python
vec1 = np.array([1,2,3])
vec2 = np.array([4,5,6])
vec3 = np.array([7, 8, 9])
 
min_max_dist(vec1, vec2, vec3)
# (5.196152422706632, 10.392304845413264)
```

In [32]:

import numpy as np

def min_max_dist(*args) -> tuple:
    """Takes any number of vectors with the same length
    and returns min and max distance between this vectors.

    Returns:
        tuple: min distance, max distance
        
    """

    args = [np.array(arg) for arg in args]
    dist =[np.linalg.norm(vec1 - vec2)
        for i, vec1 in enumerate(args[:-1]) for vec2 in args[i+1:]]
    return np.min(dist), np.max(dist)

vec1 = np.array([1,2,3])
vec2 = np.array([4,5,6])
vec3 = np.array([7, 8, 9])

min_max_dist(vec1, vec2, vec3)


(5.196152422706632, 10.392304845413264)

### Задание 10.9

Напишите функцию `any_normal`, которая принимает на вход неограниченное<br>
число векторов через запятую. Гарантируется, что все векторы, которые<br>
передаются, одинаковой длины.

Функция возвращает `True`, если есть хотя бы одна пара перпендикулярных<br>
векторов. Иначе возвращает `False`.

Пример:

```python
vec1 = np.array([2, 1])
vec2 = np.array([-1, 2])
vec3 = np.array([3,4])
print(any_normal(vec1, vec2, vec3))
# True
```

In [36]:
import numpy as np

def any_normal(*args) -> bool:
    """Takes any number of vectors with the same length
    and returns True, if at least two of them are perpendicular,
    returns False otherwise.

    Returns:
        bool: True if two of vectors are perpendicular.
    """

    args = [np.array(arg) for arg in args]
    for i, vec1 in enumerate(args[:-1]):
        for vec2 in args[i+1:]:
            if np.dot(vec1, vec2) == 0:
                return True
    return False

vec1 = np.array([2, 1])
vec2 = np.array([-1, 2])
vec3 = np.array([3,4])
print(any_normal(vec1, vec2, vec3))


True


### Задание 10.10

Напишите функцию get_loto(num), генерирующую трёхмерный массив случайных целых чисел от 1 до 100 (включительно). Это поля для игры в лото.

Трёхмерный массив должен состоять из таблиц чисел формы 5х5, то есть итоговая форма — (num, 5, 5).

Функция возвращает полученный массив.

Пример:

```python
get_loto(3)
```

Результат:

```python
array([[[ 35,  66,  38,  11,  32],
       [ 32,   7,  37,  83,  42],
       [ 89,  37,  19,  51,  89],
       [ 70, 100,  83,   5,  11],
       [ 20,  13,  60,  26,  41]],
 
      [[ 23,  49,  76,  44,  43],
       [ 59,  63,  99,  92,   2],
       [ 83,   4,  25,  73,  19],
       [ 10,  18,  40,  11,  21],
       [ 58,  45,  73,  93,  61]],
 
      [[100,  88,  70,  34,  51],
       [  5,  35,  36,  85,  88],
       [ 72,  23,  87,  30,  40],
       [ 29,  21,  51,  73,  81],
       [ 91,  19,  87,  60,  27]]])
```